openrtb_ext/request_wrapper.go

```diff
@@ -0,0 +1,773 @@
+// ---------------------------------------------------------------
+
+type AppExt struct {
+	ext         map[string]json.RawMessage
+	extDirty    bool
+	prebid      *ExtAppPrebid
+	prebidDirty bool
+}
+
+func (ae *AppExt) Unmarshal(extJson json.RawMessage) error {
+	if len(ae.ext) != 0 || ae.Dirty() {
+		return nil
+	}
+	ae.ext = make(map[string]json.RawMessage)
+	if len(extJson) == 0 {
+		return nil
+	}
+	err := json.Unmarshal(extJson, &ae.ext)
+	if err != nil {
+		return err
+	}
+	prebidJson, hasPrebid := ae.ext["prebid"]
+	if hasPrebid {
+		ae.prebid = &ExtAppPrebid{}
+		err = json.Unmarshal(prebidJson, ae.prebid)
+	}
+
+	return err
+}
+
+func (ae *AppExt) Marshal() (json.RawMessage, error) {
+	if ae.prebidDirty {
+		prebidJson, err := json.Marshal(ae.prebid)
+		if err != nil {
+			return nil, err
+		}
+		if len(prebidJson) > 0 {
+			ae.ext["prebid"] = json.RawMessage(prebidJson)
+		} else {
+			delete(ae.ext, "prebid")
+		}
+	}
+
+	rawJson, err := json.Marshal(ae.ext)
+	if err == nil {
+		ae.prebidDirty = false
+	}
+	ae.extDirty = false
+	return rawJson, err
+}
+
+func (ae *AppExt) Dirty() bool {
+	return ae.extDirty || ae.prebidDirty
+}
+
+func (ae *AppExt) GetExt() map[string]json.RawMessage {
+	ext := ae.ext
+	return ext
+}
+
+func (ae *AppExt) SetExt(ext map[string]json.RawMessage) {
+	ae.ext = ext
+	ae.extDirty = true
+}
+
+func (ae *AppExt) GetPrebid() *ExtAppPrebid {
+	if ae.prebid == nil {
+		return nil
+	}
+	prebid := *ae.prebid
+	return &prebid
+}
+
+func (ae *AppExt) SetPrebid(prebid *ExtAppPrebid) {
+	ae.prebid = prebid
+	ae.prebidDirty = true
+}
+
+// ---------------------------------------------------------------
+// RegExt provides an interface for request.regs.ext
+// ---------------------------------------------------------------
+
+type RegExt struct {
+	ext            map[string]json.RawMessage
+	extDirty       bool
+	uSPrivacy      string
+	uSPrivacyDirty bool
+}
+
+func (re *RegExt) Unmarshal(extJson json.RawMessage) error {
+	if len(re.ext) != 0 || re.Dirty() {
+		return nil
+	}
+	re.ext = make(map[string]json.RawMessage)
+	if len(extJson) == 0 {
+		return nil
+	}
+	err := json.Unmarshal(extJson, &re.ext)
+	if err != nil {
+		return err
+	}
+	uspJson, hasUsp := re.ext["us_privacy"]
+	if hasUsp {
+		err = json.Unmarshal(uspJson, &re.uSPrivacy)
+	}
+
+	return err
+}
+
+func (re *RegExt) Marshal() (json.RawMessage, error) {
+	if re.uSPrivacyDirty {
+		if len(re.uSPrivacy) > 0 {
+			rawjson, err := json.Marshal(re.uSPrivacy)
+			if err != nil {
+				return nil, err
+			}
+			re.ext["us_privacy"] = rawjson
+		} else {
+			delete(re.ext, "us_privacy")
+		}
+	}
+	if len(re.ext) == 0 {
+		re.uSPrivacyDirty = false
+		return nil, nil
+	}
+
+	rawJson, err := json.Marshal(re.ext)
+	if err == nil {
+		re.uSPrivacyDirty = false
+	}
+	return rawJson, err
+}
+
+func (re *RegExt) Dirty() bool {
+	return re.uSPrivacyDirty
+}
+
+func (re *RegExt) GetExt() map[string]json.RawMessage {
+	ext := re.ext
```

Comment:
Should we protect these pointer receivers from nil objects? What would happen if the `re` is nil?

Reply:
I discussed with @hhhjort to add defensive programming at the RequestWrapper level and then not to worry about nil in the function receivers. It would be annoying, to say the least, with having to add the nil check in all receivers and use CPU cycles to always form the check. That's what unit tests are for. :)